exporter/internal/queue/sized_elements_channel.go

```diff
@@ -0,0 +1,113 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package queue // import "go.opentelemetry.io/collector/exporter/internal/queue"
+
+import "sync/atomic"
+
+// sizedElementsChannel is a channel wrapper for sized elements with a capacity set to a total size of all the elements.
+// The channel will accept elements until the total size of the elements reaches the capacity.
+type sizedElementsChannel[T any] struct {
+	used *atomic.Int64
+
+	// We need to store the capacity in a separate field because the capacity of the channel can be higher.
+	// It happens when we restore a persistent queue from a disk that is bigger than the pre-configured capacity.
+	cap int64
+	ch  chan T
+}
+
+type sizedElementsChannelOption[T any] func(*sizedElementsChannel[T])
+
+// withPreloadElements puts the elements into the queue with the given size. It's used by the persistent queue to
+// initialize the queue with the elements recovered from the disk.
+func withPreloadElements[T any](els []T, totalSize int64) sizedElementsChannelOption[T] {
+	return func(vcq *sizedElementsChannel[T]) {
+		vcq.used.Store(totalSize)
+		capacity := int(vcq.cap)
+		if capacity < len(els) {
+			capacity = len(els)
+		}
+		vcq.ch = make(chan T, capacity)
+		for _, el := range els {
+			vcq.ch <- el
+		}
+	}
+}
+
+// newSizedElementsChannel creates a sized elements channel. Each element is assigned a size by the provided sizer.
+// chanCapacity is the capacity of the underlying channel which usually should be equal to the capacity of the queue to
+// avoid blocking the producer.
+func newSizedElementsChannel[T any](capacity int, opts ...sizedElementsChannelOption[T]) *sizedElementsChannel[T] {
+	sech := &sizedElementsChannel[T]{
+		used: &atomic.Int64{},
+		cap:  int64(capacity),
+	}
+	for _, opt := range opts {
+		opt(sech)
+	}
+	if sech.ch == nil {
+		sech.ch = make(chan T, capacity)
```

Comment:
This will be a problem if we switch to use bytes, since it will be very large and lots of memory.

Reply:
Right. We can think of a solution outside of this PR, right?